src/coreclr/vm/method.hpp

```diff
@@ -1839,19 +1851,26 @@ class MethodDesc
         if (!HasAsyncMethodData())
             return false;
 
-        // Only user runtimeasync methods operate on copies. runtime-async ->
-        // compiler-async thunks do not.
-        AsyncMethodType asyncType = GetAddrOfAsyncMethodData()->type;
+        // Only async2 methods backed by actual user code operate on copies.
+        // Thunks with runtime-supplied implementation do not.
+        AsyncMethodKind asyncType = GetAddrOfAsyncMethodData()->type;
 
-        return asyncType == AsyncMethodType::Async;
+        return asyncType == AsyncMethodKind::Async;
```

Comment:
Shouldn't this be `AsyncImplHelper`, since you changed the meaning of `Async`?

Reply:
Yes. This is a bug.